Src/IronPython/Compiler/Parser.cs

```diff
@@ -1737,12 +1766,18 @@ struct WithItem {
                     return ErrorStmt();
                 }
 
-                if (metaclass != null) {
-                    l.Add(new AssignmentStatement(new[] { new NameExpression("__metaclass__") }, metaclass));
+                if (clsContext != null) {
+                    if (!clsContext.InMember && clsContext.Metaclass != null) {
+                        l.Add(new AssignmentStatement(new[] { new NameExpression("__metaclass__") }, clsContext.Metaclass));
+                    }
+                    
+                    if (clsContext.InMember) {
+                        l.Add(new AssignmentStatement(new[] { new NameExpression("__class__") }, new NameExpression(clsContext.Name)));
```

Comment:
Looks like I need to insert an assignment only if super or \_\class\_\_ are found in the method, but only the first time it's encountered in the method.